sdk/openai/openai/sources/customizations/api/sse.ts

```diff
@@ -1,46 +1,166 @@
+            break;
+        }
       }
-      try {
-        const event = JSON.parse(prevLineIfIncomplete + str);
-        prevLineIfIncomplete = "";
-        events.push(toEvent(event));
-      } catch (e) {
-        prevLineIfIncomplete += str;
+      if (end === -1) {
+        // We reached the end of the buffer but the line hasn't ended.
+        // Wait for the next chunk and then continue parsing:
+        break;
       }
+      onLine(buf.subarray(start, end), fieldLen);
+      start = bufIdx; // we're now on the next line
+      fieldLen = -1;
     }
-    return events;
-  });
+    if (start === bufLen) {
+      buf = undefined;
+    } else if (start !== 0) {
+      // discard already processed lines
+      buf = buf.subarray(start);
+      bufIdx -= start;
+    }
+  };
 }
 
-async function* streamToEvents<T>(
-  stream: AsyncIterable<string>,
-  processChunk: (chunk: string) => T[]
-): AsyncIterable<T> {
-  for await (const chunk of stream) {
-    yield* processChunk(chunk);
+function concatBuffer(a: Uint8Array, b: Uint8Array): Uint8Array {
+  const res = new Uint8Array(a.length + b.length);
+  res.set(a);
+  res.set(b, a.length);
+  return res;
+}
+
+function getMessages(
+  onId: (id: string) => void,
+  onRetry: (intervalInMs: number) => void,
+  onMessage: (msg: EventMessage) => void
+): (line: Uint8Array, fieldLen: number) => void {
+  let message = createMessage();
+  const decoder = new TextDecoder();
+  return function onLine(line: Uint8Array, fieldLen: number) {
+    if (line.length === 0 && message.data !== undefined) {
+      // empty line denotes end of message. Trigger the callback and start a new message:
+      onMessage(message as EventMessage);
+      message = createMessage();
+    } else if (fieldLen > 0) {
+      // exclude comments and lines with no values
+      // line is of format "<field>:<value>" or "<field>: <value>"
+      // https://html.spec.whatwg.org/multipage/server-sent-events.html#event-stream-interpretation
+      const field = decoder.decode(line.subarray(0, fieldLen));
+      const valueOffset = fieldLen + (line[fieldLen + 1] === ControlChars.Space ? 2 : 1);
+      const value = decoder.decode(line.subarray(valueOffset));
+
+      switch (field) {
+        case "data":
+          message.data = message.data ? message.data + "\n" + value : value;
+          break;
+        case "event":
+          message.event = value;
+          break;
+        case "id":
+          onId((message.id = value));
+          break;
+        case "retry": {
+          const retry = parseInt(value, 10);
+          if (!isNaN(retry)) {
+            onRetry((message.retry = retry));
+          }
+          break;
+        }
+      }
+    }
+  };
+}
+
+type PartialSome<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;
+
+function createMessage(): PartialSome<EventMessage, "data"> {
+  return {
+    data: undefined,
+    event: "",
+    id: "",
+    retry: undefined,
+  };
+}
+
+export function onSSE(
```

Comment:
I think this is the main feedback here. It is a good point and I initially attempted to do it this way since it is natural but it proved to be difficult given the amount of state involved. I can give it another try.

Reply:
I can suffer with you in one of the pair rooms if you want, I believe you that it's tricky :)